modules/win_scheduled_tasks/files/azure-maintainsystem.ps1

```diff
@@ -407,47 +440,112 @@ function Check-AzVM-Name {
+## If the managed-by tag is set to taskcluster and the packer hand off is complete
 If (($hand_off_ready -eq 'yes') -and ($managed_by -eq 'taskcluster')) {
-  if (!(Test-VolumeExists -DriveLetter 'Y') -and !(Test-VolumeExists -DriveLetter 'Z')) {
+  ## We use a second data disk with 2 partitions, if that doesn't exist, mount the disk and set the drive letters
+  if (-Not (Test-VolumeExists -DriveLetter 'Y') -and -Not (Test-VolumeExists -DriveLetter 'Z')) {
     AzMount-DiskTwo
     Set-DriveLetters
   }
-  Check-AzVM-Name
+  ## Set the VM the name that taskcluster gave it, if it's not already set
+  Set-AzVMName
+  ## Clean the Z:\task_* & C:\Users\task_* directories, and any old log under C:\logs\old
   Run-MaintainSystem
   if (((Get-ItemProperty "HKLM:\SOFTWARE\Mozilla\ronin_puppet").inmutable) -eq 'false') {
     Puppet-Run
   }
-  StartWorkerRunner
+  ## Start worker runner, which starts generic-worker
+  Start-WorkerRunner
   # wait and check if GW has started
   # Followed by additional checks to ensure VM is productive if up
-  start-sleep -s 900
+  Start-Sleep -Seconds 900
+  ## if it doesn't start at this point, we need to figure out what's going on
+  ## TODO: Check worker-runner-service logs
   while($true) {
-    $gw = (Get-process -name generic-worker -ErrorAction SilentlyContinue )
-    if ($gw -eq $null) {
+    $gw = (Get-process -name "generic-worker" -ErrorAction SilentlyContinue )
+    if ($null -eq $gw) {
+      ## check to see if there are any scheduled events
+      ## https://learn.microsoft.com/en-us/azure/virtual-machines/windows/scheduled-events
+      $events = Get-AzureInstanceMetadataScheduledEvents -ApiVersion "2020-07-01"
+      ## if there are scheduled events, log it here
+      if ($null -ne $events.events) {
+        foreach ($thing in $events.events) {
+          Write-Log -Message ('{0} :: Azure VM Maintenance identified: {1}' -f $($MyInvocation.MyCommand.Name),$thing.eventtype) -severity 'DEBUG'
+        }
+        ## Start the Azure VM Maintenance event
+        ## https://learn.microsoft.com/en-us/azure/virtual-machines/windows/scheduled-events#start-an-event
+        ## There could be multiple events, so just pick the eventId of the first one
+        $scheduled_event_post = Set-AzureInstanceMetadataScheduledEvents -EventID ($events.events.EventId|Select-Object -First 1)
+        ## Check to see if it actually started
+        Write-Log -Message ('{0} :: Azure VM Maintenance started. Status code: {1}' -f $($MyInvocation.MyCommand.Name),$scheduled_event_post.status_code) -severity 'DEBUG'
+      }
       # Wait to supress meesage if check is cuaght during a reboot.
       start-sleep -s 45
-      Write-Log -message  ('{0} :: UNPRODUCTIVE: Generic-worker process not found after expected time' -f $($MyInvocation.MyCommand.Name)) -severity 'DEBUG'
+      Write-Log -message ('{0} :: UNPRODUCTIVE: Generic-worker process not found after expected time' -f $($MyInvocation.MyCommand.Name)) -severity 'DEBUG'
```

Comment:
On line 552, let's reduce that sleep to 1 second. 

All these long sleeps were left over from the initial migration. Things are running much smoother now, so the reduction should be fine.